text/0000-op-assign.md

```diff
@@ -0,0 +1,74 @@
+- Feature Name: op_assign
+- Start Date: 2015-03-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add the family of `[Op]Assign` traits to allow overloading assignment
+operations like `a += b`.
+
+# Motivation
+
+We already let users overload the binary operations, letting them overload the
+assignment version is the next logical step. Plus, this sugar is important to
+make mathematical libraries more palatable.
+
+# Detailed design
+
+Add the following **unstable** traits to libcore and reexported them in stdlib:
+
+```
+// `+=`
+#[lang = "add_assign"]
+trait AddAssign<Rhs=Self> {
+    fn add_assign(&mut self, &Rhs);
+}
+
+// the remaining traits have the same signature
+// (lang items have been omitted for brevity)
+trait BitAndAssign { .. }  // `&=`
+trait BitOrAssign { .. }   // `|=`
+trait BitXorAssign { .. }  // `^=`
+trait DivAssign { .. }     // `/=`
+trait MulAssign { .. }     // `*=`
+trait RemAssign { .. }     // `%=`
+trait ShlAssign { .. }     // `<<=`
+trait ShrAssign { .. }     // `>>=`
+trait SubAssign { .. }     // `-=`
+```
+
+Implement these traits for the primitive numeric types *without* overloading,
+i.e. only `impl AddAssign<i32> for i32 { .. }`.
+
+Add an `op_assign` feature gate. When the feature gate is enabled, the compiler
+will consider these traits when typecheking `a += b`. Without the feature gate
+the compiler will enforce that `a` and `b` must be primitives of the same
+type/category as it does today.
+
+Once we feel comfortable with the implementation we'll remove the feature gate
+and mark the traits as stable. This can be done after 1.0 as this change is
+backwards compatible.
+
+# Drawbacks
+
+None that I can think of.
+
+# Alternatives
+
+Alternatively, we could change the traits to take the RHS by value. This makes
```

Comment:
I would like to see discussion of why the proposed design differs from the plain `Add`, `Sub` etc. traits. (I don't think it's a bad idea, but it seems worthy of at least mentioning briefly.)